src/main/java/com/epam/aidial/cfg/domain/service/ExternalDeploymentService.java

```
@@ -0,0 +1,57 @@
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.networknt.schema.utils.StringUtils;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+@Service
+@LogExecution
+public class ExternalDeploymentService {
+
+    private final DeploymentClient deploymentClient;
+    private final String deploymentClientUrl;
+
+    private final Cache<UUID, DeploymentInfoDto> deploymentCache;
+
+    public ExternalDeploymentService(DeploymentClient deploymentClient,
+                                              @Value("${plugins.deployment.manager.cache.expiration.interval}") long cacheExpirationInterval,
+                                              @Value("${plugins.deployment.manager.client.url}") String deploymentClientUrl) {
+        this.deploymentClient = deploymentClient;
+        this.deploymentClientUrl = deploymentClientUrl;
+        this.deploymentCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(cacheExpirationInterval, TimeUnit.MILLISECONDS)
+                .build();
+    }
+
+    public DeploymentInfoDto getById(String id) {
+        try {
+            return deploymentCache.get(UUID.fromString(id), () -> {
+                log.debug("Deployment '{}' is not present in cache, loading from deployment client", id);
+                return getDeploymentInfoDto(id);
+            });
+        } catch (DeploymentClientNotExistsException deploymentClientNotExistsException) {
+            throw deploymentClientNotExistsException;
+        } catch (Exception e) {
+            log.error("Failed to get deployment by ID '{}'", id, e);
+            return null;
+        }
+    }
+
+    private DeploymentInfoDto getDeploymentInfoDto(String id) {
+        if (StringUtils.isBlank(deploymentClientUrl)) {
```

Could you please clarify the purpose of the custom exception?

---

Because none of existing fits. What's the problem?

---

I mean why do we need to separate case when deploymentClientUrl is not set from any other cases?

---

InterceptorImporter line 89

---

Just to log warning message "Failed to get deployment" instead of logging error message?
When `DeploymentClientNotExistsException` is thrown, `deploymentInfo` in `InterceptorImporter` stays null, when any other exception is thrown error message is logged but `deploymentInfo` in `InterceptorImporter` still stays null

---

That's important to know that deployment was expected, but wasn't found. We have plans to have pop-up warning messages on FE that will inform about issues during endpoints refresh/unsuccessful import of interceptors linked to containers